src/_xoConnection.js

```diff
@@ -0,0 +1,169 @@
+/* eslint-env jest */
+import defer from "golike-defer";
+import Xo from "xo-lib";
+import XoCollection from "xo-collection";
+import { find, forOwn } from "lodash";
+
+import config from "./_config";
+
+const logError = error => console.error(error);
+
+class XoConnection extends Xo {
+  constructor(opts) {
+    super(opts);
+
+    this._objects = new XoCollection();
+    this._watchers = {};
+    this._tempResourcesByClass = {
+      backupNg: [],
+      job: [],
+      user: [],
+    };
+  }
+
+  get objects() {
+    return this._objects;
+  }
+
+  _addObjectsListener() {
+    this.on("notification", ({ method, params }) => {
+      if (method !== "all") {
+        return;
+      }
+
+      const { _objects: objects, _watchers: watchers } = this;
+      const fn = params.type === "exit" ? objects.unset : objects.set;
+      forOwn(params.items, (item, id) => {
+        fn.call(objects, id, item);
+
+        const watcher = watchers[id];
+        if (watcher !== undefined) {
+          watcher(item);
+          delete watchers[id];
+        }
+      });
+    });
+  }
+
+  async _fetchObjects() {
+    const { _objects: objects, _watchers: watchers } = this;
+    forOwn(await this.call("xo.getAllObjects"), (object, id) => {
+      objects.set(id, object);
+
+      const watcher = watchers[id];
+      if (watcher !== undefined) {
+        watcher(object);
+        delete watchers[id];
+      }
+    });
+  }
+
+  // TODO: integrate in xo-lib.
+  waitObject(id) {
+    return new Promise(resolve => {
+      this._watchers[id] = resolve;
+    }); // FIXME: work with multiple listeners.
+  }
+
+  async getOrWaitObject(id) {
+    const object = this._objects.all[id];
+    if (object !== undefined) {
+      return object;
+    }
+    return this.waitObject(id);
+  }
+
+  @defer
+  async connect($defer, credentials = config.adminCredentials) {
+    await this.open();
+    $defer.onFailure(() => this.close());
+    await this.signIn(credentials);
+
+    this._addObjectsListener();
+    await this._fetchObjects();
+
+    return this;
+  }
+
+  async waitObjectState(id, predicate) {
+    let obj = this._objects.all[id];
+    while (true) {
+      try {
+        await predicate(obj);
+        return;
+      } catch (_) {}
+      // If failed, wait for next object state/update and retry.
+      obj = await this.waitObject(id);
+    }
+  }
+
+  async _createTempResources(method, params) {
+    const result = await super.call(method, params);
+    const [class_] = method.split(".");
+    this._tempResourcesByClass[class_].push(result.id || result);
+    return result;
+  }
```

Comment:
IMHO this abstraction is too complicated, keep things as simple as possible.

Reply:
This method mutualize the logic behind creating a temporary resource method. It will be easy to add more methods without taking care of this logic. But it can be improved. Can you please give some idea to make it better ?